### PR TITLE
Enable Download of External Resources from Artifactory Remote Repository

### PR DESCRIPTION
**Pull request by Or-Geva**
- [ ] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] I used `npm run format` for formatting the code before submitting the pull request.
-----
This PR adds the functionality to configure a remote repository within Artifactory as a proxy for essential tools that the JFrog extension typically fetches from https://releases.jfrog.io/. This configuration feature allows users to specify the externalResourcesRepository within the JFrog extension or set the JFROG_IDE_RELEASES_REPO environment variable.

Screenshot 2023-10-31 at 15 31 38

This PR includes refactoring of the JAS runner creation process and introduces a dedicated Analyzer Manager class to encapsulate all related functionalities.

### CODE REVIEW

File: src/test/tests/integration/externalResourcesRepository.test.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        assert.isDefined(expectedContent, 'Failed to read expected SecretsScanResponse content from ' + dataPath);
+        response = await runner
+            .executeRequest(() => undefined, { roots: [testDataRoot] } as AnalyzeScanRequest)
+            .then(runResult => runner.convertResponse(runResult));
+    });
+    after(() => {
+        delete process.env[Configuration.JFROG_IDE_RELEASES_REPO];
+    });
+    it('Check response defined', () => {
+        assert.isDefined(response);
```
Comment (yahavi) [inline src/test/tests/integration/externalResourcesRepository.test.ts:39]: Looks like the entire test logic is not in the test, but in the "before"

File: src/test/tests/integration/externalResourcesRepository.test.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        let dataPath: string = path.join(testDataRoot, 'expectedScanResponse.json');
+        expectedContent = JSON.parse(fs.readFileSync(dataPath, 'utf8').toString());
+        assert.isDefined(expectedContent, 'Failed to read expected SecretsScanResponse content from ' + dataPath);
+        response = await runner
+            .executeRequest(() => undefined, { roots: [testDataRoot] } as AnalyzeScanRequest)
+            .then(runResult => runner.convertResponse(runResult));
+    });
+    after(() => {
+        delete process.env[Configuration.JFROG_IDE_RELEASES_REPO];
+    });
```
Comment (yahavi) [inline src/test/tests/integration/externalResourcesRepository.test.ts:37]:
```suggestion
    });

```

File: src/test/tests/usageUtils.test.ts [hunk 3]
```diff
@@ -2,38 +2,40 @@ import { IUsageFeature } from 'jfrog-client-js';
...
 import { LogManager } from '../../main/log/logManager';
-import { EntitledScans } from '../../main/scanLogic/scanManager';
 import { PackageType } from '../../main/types/projectType';
-import { UsageUtils } from '../../main/utils/usageUtils';
+import { UsageJasScanType, UsageUtils } from '../../main/utils/usageUtils';
 import { Uri } from 'vscode';
 
 describe('Usage Utils Tests', async () => {
     const logManager: LogManager = new LogManager().activate();
+    const WithAdvanceScan: Set<UsageJasScanType> = new Set<UsageJasScanType>();
```
Comment (yahavi) [inline src/test/tests/usageUtils.test.ts:11]:
```suggestion
    const withAdvancedSecurityScan: Set<UsageJasScanType> = new Set<UsageJasScanType>();
```

File: src/test/tests/utils/MockJasRunnerFactory.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { IacRunner } from '../../../main/scanLogic/scanRunners/iacScan';
+import { SecretsRunner } from '../../../main/scanLogic/scanRunners/secretsScan';
+import { JasRunnerFactory } from '../../../main/scanLogic/sourceCodeScan/jasRunnerFactory';
+
+export class MockJasRunnerFactory extends JasRunnerFactory {
```
Comment (yahavi) [inline src/test/tests/utils/MockJasRunnerFactory.ts:5]: What's the point of this mock if it does exactly like its parent class?

**Review by yahavi — CHANGES_REQUESTED** [review]

File: src/main/scanLogic/scanRunners/jasRunner.ts [hunk 5]
```diff
@@ -127,66 +75,26 @@ export abstract class JasRunner {
...
-            console.log(text);
-        }
-        this._logManager.logMessage(text, isErr ? 'ERR' : 'DEBUG');
+        await this._analyzerManager.runWithTimeout(checkCancel, args, executionLogDirectory);
     }
 
     protected logStartScanning(request: AnalyzeScanRequest): void {
         this._logManager.logMessage(
-            `Scanning directories ' ${request.roots} + ', for ${this._scanType} issues. Skipping folders: ${request.skipped_folders}`,
+            `Scanning directories '${request.roots}', for ${this._scanType} issues. Skipping folders: ${request.skipped_folders}`,
```
Comment (yahavi) [inline src/main/scanLogic/scanRunners/jasRunner.ts:97]: In sast the "skipped_folder" is "exclude_patterns". It looks bad in the log:
```
[DEBUG - 5:17:01 PM] Scanning directories '/Users/yahavi/code/bad-guys', for sast issues. Skipping folders: undefined
[DEBUG - 5:17:01 PM] Input YAML:
scans:
  - type: sast
    roots:
      - /Users/yahavi/code/bad-guys
    exclude_patterns:
      - '**/*.venv*/**'
```

File: src/main/scanLogic/scanRunners/analyzerManager.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+     */
+    public getUrlPath(): string {
+        return Utils.addZipSuffix(
+            AnalyzerManager.RELATIVE_DOWNLOAD_URL +
+                '/' +
+                AnalyzerManager.ANALYZER_MANAGER_VERSION +
+                '/' +
+                Utils.getArchitecture() +
+                '/' +
+                AnalyzerManager.BINARY_NAME
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/scanLogic/scanRunners/analyzerManager.ts:82]: This doesn't represent the exact binary; instead, it's a zipped version of the binary. Refer to line 76 for the zipped file. The reference to the 'binary name' you're mentioning can be found on line 103.

File: src/main/scanLogic/scanRunners/analyzerManager.ts [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        );
+    }
+
+    private async checkForUpdates(): Promise<boolean> {
+        if (await this._binary.isOutdated()) {
+            this._logManager.logMessage('Updating Advanced Security Features', 'INFO');
+            try {
+                return await this._binary.update();
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/scanLogic/scanRunners/analyzerManager.ts:90]: Within the run function, environment variables with proxy settings are established for the analyzer manager. In the update flow, the jfrogClient is utilized just like all other client interactions that proceed to the same location in the code, utilizing createJfrogClient, which already manages the proxy configuration.

File: src/main/scanLogic/scanRunners/jasRunner.ts (hunk 5 above)
Comment (Or-Geva) [reply src/main/scanLogic/scanRunners/jasRunner.ts:97]: It is already been fixed at https://github.com/jfrog/jfrog-vscode-extension/pull/433/

**Pull request merged**